### ISSUE TITLE
causal_async_star_stack_presence_test flaked

### ISSUE THREAD

**Issue by floitschG** [dart-lang/sdk#28885]
The buildbot just flaked on the causal_async_star_stack_presence_test.
Please investigate it, fix it, and/or consider marking the test as flaky.

The stdout is here (and will go stale soon): https://build.chromium.org/p/client.dart/builders/vm-linux-release-ia32-optcounter-threshold-be/builds/4365/steps/checked%20vm%20tests/logs/stdio

The relevent output:
```
FAILED: none-vm-checked release_ia32 service/causal_async_star_stack_presence_test
Expected: Pass 
Actual: RuntimeError
CommandOutput[vm]:

stdout:
** Launching SERVICE_TEST_TESTEE=true out/ReleaseIA32/dart -DDART_SERVICE_USE_AUTH=false --enable_asserts --enable_type_checks --error_on_bad_type --error_on_bad_override --verbose_debug --optimization-counter-threshold=5 --ignore-unrecognized-flags --packages=/mnt/data/b/build/slave/vm-linux-release-ia32-optcounter-threshold-be/build/sdk/.packages --enable-vm-service:0 /mnt/data/b/build/slave/vm-linux-release-ia32-optcounter-threshold-be/build/sdk/runtime/observatory/tests/service/causal_async_star_stack_presence_test.dart
>testee>out> Observatory listening on http://127.0.0.1:58957/
** Signaled to run test queries on http://127.0.0.1:58957/
>testee>out> 
Loading VM...
>testee>out> HandleSteppingRequest- kStepOut f647184c
>testee>out> >>> single step break at file:///mnt/data/b/build/slave/vm-linux-release-ia32-optcounter-threshold-be/build/sdk/runtime/observatory/tests/service/causal_async_star_stack_presence_test.dart:25 (func <helper_async_body> token 99)
Done loading VM
Running causal_async_star_stack_presence_test.dart [1/11]
Paused with PauseBreakpoint
Running causal_async_star_stack_presence_test.dart [2/11]
Checking we are at line 25
Program is stopped at line: 25
Running causal_async_star_stack_presence_test.dart [3/11]
Running causal_async_star_stack_presence_test.dart [4/11]
>testee>out> helper
Running causal_async_star_stack_presence_test.dart [5/11]
>testee>out> HandleSteppingRequest- kStepOut f647183c
>testee>out> >>> single step break at file:///mnt/data/b/build/slave/vm-linux-release-ia32-optcounter-threshold-be/build/sdk/runtime/observatory/tests/service/causal_async_star_stack_presence_test.dart:18 (func <foobar_async_gen_body> token 64)
Paused with PauseBreakpoint
Running causal_async_star_stack_presence_test.dart [6/11]
Checking we are at line 18
Program is stopped at line: 18
Running causal_async_star_stack_presence_test.dart [7/11]
Running causal_async_star_stack_presence_test.dart [8/11]
Running causal_async_star_stack_presence_test.dart [9/11]
>testee>err> Dumping native stack trace for thread 75a5
>testee>err>   [0x085cb10f] dart::Zone::PrintJSON(dart::JSONStream*) const
>testee>err>   [0x085cb10f] dart::Zone::PrintJSON(dart::JSONStream*) const
>testee>err>   [0x085b852c] dart::Thread::PrintJSON(dart::JSONStream*) const
>testee>err>   [0x085bce3c] dart::ThreadRegistry::PrintJSON(dart::JSONStream*) const
>testee>err>   [0x0840fc77] dart::Isolate::PrintJSON(dart::JSONStream*, bool)
>testee>err>   [0x08583b73] Unknown symbol
>testee>err>   [0x0858c64e] dart::Service::InvokeMethod(dart::Isolate*, dart::Array const&, bool)
>testee>err>   [0x0858ca71] dart::Service::HandleIsolateMessage(dart::Isolate*, dart::Array const&)
>testee>err>   [0x0841561f] dart::IsolateMessageHandler::HandleMessage(dart::Message*)
>testee>err>   [0x0843722f] dart::MessageHandler::HandleMessages(dart::MonitorLocker*, bool, bool)
>testee>err>   [0x084373fa] dart::MessageHandler::HandleOOBMessages()
>testee>err>   [0x085b8fb1] dart::Thread::HandleInterrupts()
>testee>err>   [0x08601e0e] dart::DRT_StackOverflow(dart::NativeArguments)
>testee>err>   [0xf723f387] [Stub] CallToRuntime
>testee>err>   [0xf7068cb7] _AsyncStarStreamController.scheduleGenerator
>testee>err>   [0xf706ea5d] _AsyncStarStreamController.add
>testee>err>   [0xf706e4d8] foobar
>testee>err>   [0xf6bf87f6] _Closure.call
>testee>err>   [0xf706daf7] _AsyncStarStreamController.runBody
>testee>err>   [0xf706da0a] _AsyncStarStreamController.runBody
>testee>err>   [0xf6bf87f6] _Closure.call
>testee>err>   [0xf7064a10] _microtaskLoop
>testee>err>   [0xf7064662] _startMicrotaskLoop
>testee>err>   [0xf7064582] _startMicrotaskLoop
>testee>err>   [0xf723fa00] [Stub] DeoptimizeLazyFromReturn
>testee>err>   [0xf70644a8] _runPendingImmediateCallback
>testee>err>   [0xf6bffed2] _RawReceivePortImpl._handleMessage
>testee>err>   [0xf723f684] [Stub] InvokeDartCode
>testee>err>   [0x08324170] dart::DartEntry::InvokeFunction(dart::Function const&, dart::Array const&, dart::Array const&, unsigned int)
>testee>err>   [0x08324f1d] dart::DartEntry::InvokeFunction(dart::Function const&, dart::Array const&)
>testee>err>   [0x08327a3f] dart::DartLibraryCalls::HandleMessage(dart::Object const&, dart::Instance const&)
>testee>err>   [0x0841537a] dart::IsolateMessageHandler::HandleMessage(dart::Message*)
>testee>err>   [0x0843722f] dart::MessageHandler::HandleMessages(dart::MonitorLocker*, bool, bool)
>testee>err>   [0x0843770a] dart::MessageHandler::TaskCallback()
>testee>err> -- End of DumpStackTrace
** Killing script
Unexpected exception in service tests: NetworkRpcException(Canceling request: WebSocket closed(id: 25 method: getIsolate params: {isolateId: isolates/386575552}))
null


stderr:
Unhandled exception:
NetworkRpcException(Canceling request: WebSocket closed(id: 25 method: getIsolate params: {isolateId: isolates/386575552}))
#0      _ServiceTesterRunner.run.<anonymous closure> (file:///mnt/data/b/build/slave/vm-linux-release-ia32-optcounter-threshold-be/build/sdk/runtime/observatory/tests/service/test_helper.dart:339:9)
<asynchronous suspension>
#1      _rootRunBinary (dart:async/zone.dart:1174)
#2      _RootZone.runBinary (dart:async/zone.dart:1411)
#3      runZoned.<anonymous closure> (dart:async/zone.dart:1478)
#4      _CustomZone.handleUncaughtError (dart:async/zone.dart:1005)
#5      _Future._propagateToListeners (dart:async/future_impl.dart:552)
#6      _Future._completeError (dart:async/future_impl.dart:485)
#7      _Future._asyncCompleteError.<anonymous closure> (dart:async/future_impl.dart:537)
#8      _rootRun (dart:async/zone.dart:1150)
#9      _CustomZone.run (dart:async/zone.dart:1026)
#10     _CustomZone.runGuarded (dart:async/zone.dart:924)
#11     _CustomZone.bindCallback.<anonymous closure> (dart:async/zone.dart:951)
#12     _microtaskLoop (dart:async/schedule_microtask.dart:41)
#13     _startMicrotaskLoop (dart:async/schedule_microtask.dart:50)
#14     _runPendingImmediateCallback (dart:isolate-patch/isolate_patch.dart:96)
#15     _RawReceivePortImpl._handleMessage (dart:isolate-patch/isolate_patch.dart:149)


Command[vm]: DART_CONFIGURATION=ReleaseIA32 out/ReleaseIA32/dart --enable_asserts --enable_type_checks --error_on_bad_type --error_on_bad_override --verbose_debug --optimization-counter-threshold=5 --ignore-unrecognized-flags --packages=/mnt/data/b/build/slave/vm-linux-release-ia32-optcounter-threshold-be/build/sdk/.packages /mnt/data/b/build/slave/vm-linux-release-ia32-optcounter-threshold-be/build/sdk/runtime/observatory/tests/service/causal_async_star_stack_presence_test.dart
Took 0:00:04.700238

Short reproduction command (experimental):
    python tools/test.py -mrelease -aia32 --checked -t60 --vm-options --optimization-counter-threshold=5 service/causal_async_star_stack_presence_test
```

**Comment by johnmccutchan** [comment]
@bkonyi This looks like the VM crashed while printing a Zone to JSON.

**Comment by bkonyi** [comment]
It looks like zones sometimes fall out of scope in one thread while another thread is trying to collect information about it, causing a segfault. I've temporarily disabled the collection of information for individual zones until @johnmccutchan is back next week to discuss a different method of surfacing zone information on a per thread basis.

**Comment by bkonyi** [comment]
This was fixed in commit c0511ad2ef920518513b986e16735bf9c158aeb8.
